File: malik/ch5/orderedLinkedList.cpp
#include "orderedLinkedList.h"
#include "nodeType.cpp"
#include <iostream>

template <class Type>
bool orderedLinkedList<Type>::search(const Type& item) const
{
    bool found = false;
    nodeType<Type> *current;

    current = this->first;

    while (current != NULL && !found) {
        if (current->info >= item) found = true;
        current = current->link;
    }
    return (found) ? (current->info == item) : found;
}

template <class Type>
void orderedLinkedList<Type>::insert(const Type& item)
{
    nodeType<Type> *current;
    nodeType<Type> *trailCurrent;
    nodeType<Type> *newNode;

    bool found;

    newNode = new nodeType<Type>;
    newNode->info = item;
    newNode->link = NULL;

    if (this->first == NULL)
    {
        this->first = newNode;
        this->last = newNode;
        this->count++;
    } else {
        current = this->first;
        found = false;

        while (current != NULL && !found)
        {
            if (current->info >= item) found = true;
            else {
                trailCurrent = current;
                current = current->link;
            }
        }

        if (current == this->first)
        {
            newNode->link = this->first;
            this->first = newNode;
            this->count++;
        } else
        {
            trailCurrent->link = newNode;
            newNode->link = current;

            if (current == NULL) this->last = newNode;
            this->count++;
        }
    }
}

template <class Type>
void orderedLinkedList<Type>::insertFirst(const Type& item)
{
    insert(item);
}

template <class Type>
void orderedLinkedList<Type>::insertLast(const Type& item)
{
    insert(item);
}

template <class Type>
void orderedLinkedList<Type>::deleteNode(const Type &item)
{
    nodeType<Type> *current;
    nodeType<Type> *trailCurrent;
    bool found;

    if (this->first == NULL) std::cout << "Cannot delete from an empty list." << std::endl;
    else
    {
        current = this->first;
        found = false;

        while (current != NULL && !found)
            if (current->info >= item) found = true;
            else
            {
                trailCurrent = current;
                current = current->link;
            }

        if (current == NULL) std::cout << "The item to be deleted is not in the list." << std::endl;
        else if (current->info == item)
        {
            if (this->first == current)
            {
                this->first = this->first->link;
                if (this->first == NULL) this->last = NULL;
                delete current;
            } else
            {
                trailCurrent->link = current->link;

                if (current == this->last) this->last = trailCurrent;
                delete current;
            }
            this->count--;
        } else std::cout << "The item to be deleted is not in the list." << std::endl;
    }
}

File: malik-js/ch5/OrderedLinkedList.js
const Node = require('./Node')
const LinkedList = require('./LinkedList')

class OrderedLinkedList extends LinkedList {
    constructor() {
        super()
    }

    search(item) {
        let found = false
        let current = this._first

        while (current !== null && found === false)
            if (current.info >= item) found = true
            else current = current.link

        if (found) found = (current.info === item)

        return found
    }

    insert(item) {
        let current = new Node()
        let prev = new Node()
        let newNode = new Node(item)
        let found = false

        if (this._first === null) {
            this._first = newNode
            this._last = newNode
            this._count++
        } else {
            current = this._first
            while (current !== null && found === false)
                if (current.info >= item) found = true
                else {
                    prev = current
                    current = current.link
                }

            if (current === this._first) {
                newNode.link = this._first
                this._first = newNode
                this._count++
            } else {
                prev.link = newNode
                newNode.link = current
                if (current === null) this._last = newNode
                this._count++
            }
        }
    }

    insertFirst(item) {
        this.insert(item)
    }

    insertLast(item) {
        this.insert(item)
    }

    delete(item) {
        let current = null
        let prev = null
        let found = false
        let deleted = false

        if (this._first === null) deleted = false
        else {
            current = this._first

            while (current !== null && found === false)
                if (current.info >= item) found = true
                else {
                    prev = current
                    current = current.link
                }

            if (current === null) deleted = false
            else if (current.info === item) {
                if (this._first === current) {
                    this._first = this._first.link
                    if (this._first === null) this._last = null
                } else {
                    prev.link = current.link
                    if (current === this._last) this._last = prev
                }
                this._count--
                deleted = true
            } else {
                deleted = false
            }
        }
        return deleted
    }
}

module.exports = OrderedLinkedList
